Create a Slack command integration

The objective is to enable our company's employees to report their working hours via simple Slack command.

The backend for reporting working hours is an [Anuko Time Tracker](https://timetracker.anuko.com) that is installed on our premises. For the development process I have created a test account in the hosted version of Anuko.
url: https://timetracker.anuko.com/login.php
username: ttt-slack-staging
password: 12345678
You'll find a single project, and two tasks that you can use to report time with.

The Anuko system is a rather old, API-less system, so we have built a proxy server that enables sending time tracking commands to it. For the development process I have deployed such a proxy server that runs against the Anuko hosted version.
url: https://ttt-proxy-staging.herokuapp.com/

the proxy server has a single command -- `/track-time` -- and here's how to use it:
```
curl -X POST -d "login=ttt-slack-staging" -d "password=12345678" -d "project=ttt-slack" -d "task=consulting" -d "start=09:00" -d "finish=10:00" -d "date=2016-03-12" -d "note=hello" https://ttt-proxy-staging.herokuapp.com/track-time
```

On success, the proxy will return http 200, otherwise it returns 500 with an error message.

Your task is to implement these features:

`/ttt username {username}` to save the users' Anuko username in the database
`/ttt password {username}` to save the users' Anuko password in the database
`/ttt project {project-name}` to save the users' default project in the database
`/ttt task {task-name}` to save the users' default task in the database
`/ttt in` to be used by the user to mark a beginning of a day. Save the current time in the database
`/ttt out` to post a time tracking to Anuko, based on all the information that is already saved in the database, and the current date.
`/ttt track {project-name} {task-name} {in-time} {out-time}` for cases where the default project and task are not to be used, or for any other reason, for example the user forgot to `/ttt in` when they entered their office in the morning :)

Of course the unique username you should use for identifying users will be the one provided by Slack. Please always respond an ephemeral human readable response. For example, After a successful saving of a default project, let the user know it was successful. If there was a problem with tracking the working hours, the employee should be alerted about it. 

Please make the proxy server endpoint configurable via ENV vars, i.e. `process.env.TTT_PROXY_URI`

codemill$60

After some hours to fail start application on my computer (windows) i have donwload application on my own server but always same message : 

`{ [Error: Cannot find module '../build/Release/bson'] code: 'MODULE_NOT_FOUND' }
js-bson: Failed to load c++ bson extension, using pure JS version`

--- a/app.js
+++ b/app.js
@@ -1,3 +1,4 @@
+var assert = require('assert');
 var express = require('express');
 var path = require('path');
 var favicon = require('serve-favicon');
@@ -59,5 +60,13 @@ app.use(function(err, req, res, next) {
 // setup mongodb/monk
 app.set('dbURI', process.env.MONGOLAB_URI || 'localhost:27017/ttt-slack');
 app.set('db', require('monk')(app.get('dbURI')));
+
+// post to the anuko proxy
+assert(process.env.TTT_PROXY_URI != null, 'must set TTT_PROXY_URI in the environment');
+app.set('ttt-proxy-uri', process.env.TTT_PROXY_URI);
+
+// configure the slack token
+assert(process.env.SLACK_TOKEN != null, 'must set SLACK_TOKEN in the environment');
+app.set('slack-token', process.env.SLACK_TOKEN);
 
 module.exports = app;

--- a/package.json
+++ b/package.json
@@ -11,9 +11,11 @@
     "debug": "~2.2.0",
     "ejs": "~2.3.3",
     "express": "~4.13.1",
+    "moment": "^2.12.0",
     "mongodb": "^1.4.40",
     "monk": "^1.0.1",
     "morgan": "~1.6.1",
+    "request": "^2.69.0",
     "serve-favicon": "~2.3.0"
   }
 }

--- a/routes/slack.js
+++ b/routes/slack.js
@@ -1,6 +1,11 @@
 var express = require('express');
 var router = express.Router();
+var util = require('util');
+var moment = require('moment');
+var request = require('request');
 router.preferredBase = '/slack';
+
+const timeForms = [ 'hh:mmA', 'hh:mm', 'hA', 'hhA', 'HH', 'H:mm', 'HH:mm' ];
 
 /* GET users listing. */
 router.get('/', function(req, res, next) {
@@ -11,6 +16,253 @@ router.get('/', function(req, res, next) {
     }
     res.send(docs);
   });
+});
+
+// not-null-and-not-undefined check (I'm used to coffeescript)
+function some(x) { return x != null; }
+
+function isCommand(command) {
+  return [ 'username', 'password', 'project', 'task', 'in', 'out', 'track', 'note', 'status' ].indexOf(command) !== -1;
+}
+
+function isUpdate(command) { return !(command === 'out' || command === 'track'); }
+
+function userStatus(data) {
+  if (some(data.in)) {
+    var start = moment(data.in);
+    data.in = util.format('%s (%s)', start.format(timeForms[0]), start.fromNow());
+  }
+
+  var out = util.format(
+    'username: *%s*, password: *%s*, project: *%s*, task: *%s*, started: *%s*',
+    data.username, data.password, data.project, data.task, data.in
+  );
+
+  if (some(data.note)) {
+    out += util.format('\nnote: *%s*', data.note);
+  }
+
+  return out;
+}
+
+/* curl -X POST \
+**   -d "login=ttt-slack-staging" \
+**   -d "password=12345678" \
+**   -d "project=ttt-slack" \
+**   -d "task=consulting" \
+**   -d "start=09:00" \
+**   -d "finish=10:00" \
+**   -d "date=2016-03-12" \
+**   -d "note=hello" \
+**   https://ttt-proxy-staging.herokuapp.com/track-time */
+function mongoToAnuko(data) {
+  var tmp = {};
+  var now = moment(data.out);
+
+  tmp.login    = data.username;
+  tmp.password = data.password;
+  tmp.project  = data.project;
+  tmp.task     = data.task;
+  tmp.start    = moment(data.in).format('HH:mm');
+  tmp.finish   = now.format('HH:mm');
+  tmp.date     = now.format('YYYY-MM-DD');
+
+  if (some(data.note)) {
+    tmp.note = data.note;
+  }
+
+  return tmp;
+}
+
+function sendf(res) {
+  var args = Array.prototype.slice.call(arguments, 1);
+
+  res.send(util.format.apply(this, args));
+}
+
+// used to identify the property that is missing
+function readyForAnuko(data) {
+  var missing = null;
+  [ 'username', 'password', 'project', 'task', 'in' ].some(function (prop) {
+    if (!some(data[prop])) {
+      missing = prop;
+      return true
+    }
+  });
+
+  return missing;
+}
+
+router.post('/', function(req, res, next) {
+  // skip/ignore POST if it doesn't have Slack's API token
+  if (req.app.get('slack-token') !== req.body.token) {
+    res.status(403).send('Slack imposter!');
+    return;
+  }
+
+  // we need user_id for Mongo
+  if (!some(req.body.user_id)) {
+    res.status(400);
+    res.send('Who are you? No user_id sent from Slack.');
+    return;
+  }
+
+  const id    = req.body.user_id;
+  const users = req.app.get('db').get('users');
+
+  var args = req.body.text.split(/\s+/);
+  var command = args.shift();
+
+  if (some(command)) {
+    command = command.toLowerCase();
+  }
+
+  if (!isCommand(command)) {
+    res.status(400);
+    sendf(res, 'unrecognized command: *%s*', command);
+    return;
+  }
+
+  if (isUpdate(command)) {
+    if (args.length === 0) {
+      switch (command) {
+        case 'in':
+          args = [ new Date ];
+          break;
+        case 'status':
+          // an update-anyway query
+          args = [ id ];
+          break;
+        default:
+          res.status(400);
+          sendf(res, 'missing argument to *%s*', command);
+          return;
+      }
+    }
+
+    if (command === 'note') {
+      var tmp = /\w+\s+(.*)/.exec(req.body.text);
+      args = [ tmp[0] ]; // the note
+    }
+
+    var set = {};
+
+    // because precomputed properties :(
+    switch (command) {
+      case 'username': set.username = args[0]; break;
+      case 'password': set.password = args[0]; break;
+      case 'in'      : set.in       = args[0]; break;
+      case 'project' : set.project  = args[0]; break;
+      case 'task'    : set.task     = args[0]; break;
+      case 'note'    : set.note     = args[0]; break;
+      case 'status'  : set.id       = args[0]; break;
+    }
+
+    users.findAndModify(
+      { id: id }, // query
+      { $set: set }, // update
+      { new: true, upsert: true }, // return most-current, insert if non-existent
+      function (err, data) {
+        if (some(err)) {
+          res.status(500);
+          sendf(res, 'MongoDB error: *%s*', err);
+          return;
+        }
+
+        res.status(200).send(userStatus(data));
+      }
+    );
+    return;
+  }
+
+  // if we're here, we're trying to POST to Anuko (/ttt out & /ttt track ...)
+  users.findOne(
+    { id: id },
+    function (err, data) {
+      if (some(err)) {
+        res.status(500);
+        sendf(res, 'DB: *%s*', err);
+        return;
+      }
+
+      if (!some(data)) {
+        data = {};
+      }
+
+      if (command === 'track') {
+        if (args.length < 4) {
+          res.status(400).send('not enough arguments: `/ttt track &lt;project&gt; &lt;task&gt; &lt;start-time&gt; &lt;finish-time&gt; [&lt;note&gt;]`');
+          return;
+        }
+
+
+        args[2] = moment(args[2], timeForms, true);
+
+        if (!args[2].isValid()) {
+          res.status(400);
+          sendf(res, '`start-time` was in the wrong format (example: `%s`)', moment().format(timeForms[0]));
+          return;
+        }
+
+        args[2] = args[2].toDate();
+
+        args[3] = moment(args[3], timeForms, true);
+
+        if (!args[3].isValid()) {
+          res.status(400);
+          sendf(res, '`finish-time` was in the wrong format (example: `%s`)', moment().format(timeForms[0]));
+          return;
+        }
+
+        args[3] = args[3].toDate();
+
+
+        args.splice(4); // remove the split/mangled note
+        var tmp = /\w+\s+\w+\s+\w+\s+\w+\s+(.*)/.exec(req.body.text);
+        if (some(tmp)) {
+          args.push(tmp[1]); // the note
+        }
+
+        [ 'project', 'task', 'in', 'out', 'note' ].forEach(function (p) {
+          data[p] = args.shift();
+        });
+      }
+
+      var missing = readyForAnuko(data);
+
+      if (some(missing)) {
+        res.status(428);
+        if (missing === 'in') {
+          res.send('first do: `/ttt in`');
+        }
+        else {
+          sendf(res, 'first do: `/ttt %s &lt;%s&gt;`', missing, missing);
+        }
+        return;
+      }
+
+      var uri  = req.app.get('ttt-proxy-uri');
+      var form = mongoToAnuko(data);
+
+      res.status(200);
+      sendf(res, 'registering *%s* task with Amuko, try again if not successful in 5 seconds', data.task);
+
+      request.post({ url: uri, form: form }, function (err, post, body) {
+        var ack = 'unsuccessful - please try again';
+
+        if (!some(err) && post.statusCode === 200) {
+          ack = util.format('task *%s* has been recorded in Amuko! (%s) :smile:', data.task, body || err);
+
+          // clear the task & start-time fields
+          users.findAndModify({ id: id }, { $unset: { in: true, note: true } });
+        }
+
+        // even the POST back to Slack can fail - this is a one-off, hit-or-miss acknowledgement
+        request.post({ url: req.body.response_url, json: true, body: { text: ack } });
+      });
+
+    }
+  );
 });
 
 module.exports = router;

